src/utils/pgwire/src/pg_protocol.rs

```diff
@@ -199,37 +208,122 @@ where
+
+        // Processing the message itself.
+        //
+        // Note: pin the future to avoid stack overflow as we'll wrap it multiple times
+        // in the following code.
+        let fut = Box::pin(self.do_process_inner(msg));
+
+        // Set the current session as the context when processing the message, if exists.
+        let fut = async move {
             if let Some(session) = weak_session {
-                Either::Left(CURRENT_SESSION.scope(session, fut))
+                CURRENT_SESSION.scope(session, fut).await
             } else {
-                Either::Right(fut)
+                fut.await
             }
         };
 
-        let result = AssertUnwindSafe(fut)
-            .rw_catch_unwind()
-            .await
-            .unwrap_or_else(|payload| {
-                Err(PsqlError::Panic(
-                    panic_message::panic_message(&payload).to_owned(),
-                ))
-            })
-            .inspect_err(|error| error!(error = %error.as_report(), "error when process message"));
+        // Catch unwind.
+        let fut = async move {
+            AssertUnwindSafe(fut)
+                .rw_catch_unwind()
+                .await
+                .unwrap_or_else(|payload| {
+                    Err(PsqlError::Panic(
+                        panic_message::panic_message(&payload).to_owned(),
+                    ))
+                })
+        };
+
+        // Slow query log.
+        let fut = async move {
+            let period = *SLOW_QUERY_LOG_PERIOD;
+            let mut fut = std::pin::pin!(fut);
+            let mut elapsed = Duration::ZERO;
+
+            // Report the SQL in the log periodically if the query is slow.
+            loop {
+                match tokio::time::timeout(period, &mut fut).await {
+                    Ok(result) => break result,
+                    Err(_) => {
+                        elapsed += period;
+                        tracing::info!(
+                            target: PGWIRE_SLOW_QUERY_LOG,
+                            ?elapsed,
+                            "slow query"
+                        );
+                    }
+                }
+            }
+        };
+
+        // Query log.
+        let fut = async move {
+            let start = Instant::now();
+            let result = fut.await;
+            let elapsed = start.elapsed();
+
+            // Always log if an error occurs.
+            if let Err(error) = &result {
+                tracing::error!(error = %error.as_report(), "error when process message");
+            }
 
-        match result {
+            // Log to optionally-enabled target `PGWIRE_QUERY_LOG`.
+            // Only log if we're currently in a tracing span set in `span_for_msg`.
+            if !tracing::Span::current().is_none() {
+                tracing::info!(
+                    target: PGWIRE_QUERY_LOG,
+                    status = if result.is_ok() { "ok" } else { "err" },
+                    time = ?elapsed,
```

Comment:
BTW, shall we print the `elapsed` in a fixed width and fixed unit format?

Reply:
Can you elaborate more? I'm afraid it's hard to do this for the duration from `50 µs = 0.05 ms` to `5 min = 300 000 ms`. 🥺

Reply:
I think `ms` is enough for batch query. How about using `.as_millis()`? With the same time unit could let us easy to grep these lines.